### ISSUE TITLE
Start-Process does not populate the process object's exit code with NoNewWindow switch

### ISSUE THREAD

**Issue by unlikelier** [PowerShell/PowerShell#20400]
### Prerequisites

- [X] Write a descriptive title.
- [x] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I noticed that the `Start-Process` cmdlet sometimes does not populate the `ExitCode` property in the output, I believe it should.
This happed at least when `-PassThru`, `-Wait`, and `-NoNewWindow` switches are provided. `-NoNewWindow` seems to be the reason, since the exit code is available without it.

## Steps to reproduce
You can see this behavior with this simple test script
```
Set-Content -Value "exit 1" -Path script.ps1
$p = start-process "pwsh" script.ps1 -PassThru -NoNewWindow -Wait
$p.ExitCode  # Should output: 1
$p = start-process "pwsh" script.ps1 -PassThru -Wait
$p.ExitCode # Outputs: 1
```

### Expected behavior

```console
PS C:\> set-content -Value "exit 1" -Path script.ps1
PS C:\> $p = start-process "pwsh" script.ps1 -PassThru -NoNewWindow -Wait
PS C:\> $p.ExitCode
1
PS C:\> $p = start-process "pwsh" script.ps1 -PassThru -Wait
PS C:\> $p.ExitCode
1
```

### Actual behavior

```console
PS C:\> set-content -Value "exit 1" -Path script.ps1
PS C:\> $p = start-process "pwsh" script.ps1 -PassThru -NoNewWindow -Wait
PS C:\> $p.ExitCode
PS C:\> $p = start-process "pwsh" script.ps1 -PassThru -Wait
PS C:\> $p.ExitCode
1
```

### Environment data

```powershell
PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.0-preview.5
PSEdition                      Core
GitCommitId                    7.4.0-preview.5
OS                             Microsoft Windows 10.0.20348
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
Although the symptoms have changed a bit,  the root cause is likely the same as in:

* https://github.com/PowerShell/PowerShell/issues/5421#issuecomment-416354272

That is, it seems that an exception occurs behind the scenes due to the process handle not being available (for reasons unknown to me), which PowerShell quietly swallows, causing `$null` to be returned from `.ExitCode` (and `.Handle`).

This now seems to _always_ happen with `-NoNewWindow`, whereas it previously didn't happen if `-Wait` was specified too.

  That is, both `($p = start-process "pwsh" script.ps1 -PassThru -NoNewWindow -Wait).Handle` and `$p = start-process "pwsh" script.ps1 -PassThru -NoNewWindow; $p.WaitForExit(); $p.Handle` now output nothing, and therefore also return `$null` via `.ExitCode`.

In Windows PowerShell (and seemingly also earlier versions of PS Core), the `-Wait` variant (only) _does_ work.

**Comment by shoddyguard** [comment]
We've just hit this too (as we have PowerShell Core set to auto-update).
Unfortunately we use `Start-Process` in a lot of places so it's going to be quite painful to implement the above workaround 😞

**Comment by prayaas-a** [comment]
+1

**Comment by alexvergilis** [comment]
+1

**Comment by chrispy81** [comment]
+1

**Comment by jlpetz** [comment]
what the heck, this is pretty bad. How did this slip past testing. I have a number of scripts using '-wait' which are now all bonkers because the ExitCode and various other properties on system.diagnotic.process objects are now missing since 7.4 update. Can we please fix this ASAP

**Comment by jlpetz** [comment]
Weird, is this as known issue too?

In one of my scripts I tried swapping from in Start-Process

```
$runresult = Start-Process "$initExe" -ArgumentList "$initExeArgs" -wait -NoNewWindow -PassThru
if ($runresult.exitcode -ne 0)
```

to a simple call operator ('&'), thinking this would do the trick
```
& "$initExe" $initExeArgs
```

And while it works in a Terminal prompt, where the EXE sends prompt output to the terminal and then waits for input (as below)
```
Enter password for your user: 
```

If I run that same '&' call embedded in a .ps1 script. It seems to have the Input/Output redirection issues, as the saem EXE in the same terminal doesn't output the prompt text. Instead it just hangs waiting for input. It still works, but it's confusing as to why the prompt goes missing in that context?

**Comment by jlpetz** [comment]
Ok, got it working similar to what @mklement0 posted in https://github.com/PowerShell/PowerShell/issues/20400#issuecomment-1740954070

The one difference is that to get the same behaviour in terms of screen output when running manually in a terminal AND when running it as part of a larger .ps1, I merged had to merge the WaitForExit() line. If I didn't do this, then the Output (input prompt) from the EXE was showing up on a PS prompt line (with the current directory prefix) when it was run in a terminal (but not when run in a large .ps1). 

By merging these onto a single line for some reason it then shows the EXE prompt goes to a new line (consistently with running the same command in a larger .ps1)

So in the end I went from
```
$runresult = Start-Process "$init" -ArgumentList "$initOptions" -wait -NoNewWindow -PassThru
if ($runresult.exitcode -ne 0)
```

to 
```
$runresult = [System.Diagnostics.Process]::Start('pwsh', "-command & '$init' $initoptions") ; $runresult.WaitForExit()
if ($runresult.ExitCode -ne 0)
```

**Comment by softwarebear** [comment]
Just to add ... this doesn't happen on MacOS ...

```
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Darwin 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
@softwarebear, I see it on macOS too (same PS and OS version); e.g. 
`(Start-Process -Wait -NoNewWindow date).ExitCode`
outputs just the date and no exit code.

**Comment by mklement0** [comment]
Good to know, @softwarebear. 
Similarly, Linux seems to be unaffected (simple test command, for all platforms:
`(Start-Process -Wait -NoNewWindow -PassThru whoami).ExitCode`
Should print both the username and exit code `0`)

**Comment by mklement0** [comment]
@jlpetz, @derkveenhof found a simpler workaround - see [my updated comment](https://github.com/PowerShell/PowerShell/issues/20400#issuecomment-1740954070) above.

**Comment by softwarebear** [comment]
> Just to add ... this doesn't happen on MacOS ...
> 
> ```
> Name                           Value
> ----                           -----
> PSVersion                      7.4.0
> PSEdition                      Core
> GitCommitId                    7.4.0
> OS                             Darwin 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000
> Platform                       Unix
> PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
> PSRemotingProtocolVersion      2.3
> SerializationVersion           1.1.0.1
> WSManStackVersion              3.0
> ```

Just to be consistent ...

`
PS /Users/softwarebear/dev/1e/taas> (Start-Process -Wait -NoNewWindow -PassThru whoami).ExitCode                                                     softwarebear                    
0
PS /Users/softwarebear/dev/1e/taas> (Start-Process -Wait -PassThru whoami).ExitCode                
softwarebear                    
0
PS /Users/softwarebear/dev/1e/taas> 
`

**Comment by mklement0** [comment]
Thanks. On Unix-like platforms `-NoNewWindow` is _implied_, that's why both commands behave the same there.
On _Windows_ - and that is the bug - only the form _without_ `-NoNewWindow` reports an exit code (while - as expected - showing no `whoami.exe` output _in the calling console_, because it runs in a separate window).

**Comment by harmen-xb** [comment]
I am also experiencing this issue now after upgrading to PowerShell v7.4.0 x64 on Windows.

In PowerShell v7.3.10 it works fine:
![image](https://github.com/PowerShell/PowerShell/assets/97173058/97b42b06-6fd5-41b6-8306-c146d76ce602)

In PowerShell v7.4.0 it doesn't return the ExitCode with -NoNewWindow, while it does return the ExitCode without -NoNewWindow:
![image](https://github.com/PowerShell/PowerShell/assets/97173058/61d3e3a4-59a1-4af7-b64c-70bfa2d2b3d4)

**Comment by rkieslinger** [comment]
We have the same issue on Windows.

**Comment by santhonisz** [comment]
+1 on Windows Server 2022 and Windows 11

**Comment by mklement0** [comment]
It looks like this will be fixed in v7.4.1:

* #20866
